src/translatador.cpp

```diff
@@ -0,0 +1,385 @@
+    try {
+        return function();
+    } catch (std::exception& e) {
+        last_error = std::string(e.what());
+        return nullptr;
+    }
+}
+
+int run_fallible(const std::function<void()>& function) {
+    try {
+        function();
+        return TRL_OK;
+    } catch (std::exception& e) {
+        last_error = std::string(e.what());
+        return TRL_ERROR;
+    }
+}
+
+std::shared_ptr<marian::Options> parse_options(const char* yaml) {
+    std::shared_ptr<marian::Options> options = std::make_shared<marian::Options>();
+
+    const marian::ConfigParser parser(marian::cli::mode::translation);
+    options->merge(parser.getConfig());
+
+    // Default properties, based on ones used for Firefox translation models: https://github.com/mozilla/firefox-translations-models/blob/main/evals/translators/bergamot.config.yml
+    options->set<size_t>("max-length-break", 128);
+    options->set<float>("max-length-factor", 2.0);
+    options->set<size_t>("beam-size", 1);
+    options->set<float>("normalize", 1.0);
+    options->set<float>("word-penalty", 0.0);
+    options->set<bool>("skip-cost", true);
+    options->set<size_t>("workspace", 128);
+    options->set<std::string>("alignment", "soft");
+    options->set<std::string>("ssplit-mode", "paragraph");
+    options->set<std::string>("gemm-precision", "int8shiftAlphaAll");
+    options->set<bool>("quiet", true);
+    options->set<bool>("quiet-translation", true);
+
+    if (yaml) {
+        options->parse(std::string(yaml));
+    }
+
+    // Dummy values, should not be overridden
+    options->set<std::vector<std::string>>("vocabs", {"source", "target"});
+
+    return options;
+}
+
+TrlModel instantiate_model(const std::shared_ptr<ModelData>& data) {
+    const marian::DeviceId device(0, marian::DeviceType::cpu);
+    std::shared_ptr<marian::ExpressionGraph> graph = std::make_shared<marian::ExpressionGraph>(true);
+    graph->setDefaultElementType(marian::typeFromString(data->options->get<std::vector<std::string>>("precision", {"float32"})[0]));
+    graph->setDevice(device);
+    graph->getBackend()->configureDevice(data->options);
+    graph->reserveWorkspaceMB(data->options->get<size_t>("workspace"));
+
+    std::vector<std::shared_ptr<marian::Scorer>> scorers = marian::createScorers(data->options, std::vector<const void *>{data->model_memory.begin()});
+    for (const std::shared_ptr<marian::Scorer>& scorer : scorers) {
+        scorer->init(graph);
+        if (data->short_list_generator) {
+            scorer->setShortlistGenerator(data->short_list_generator);
+        }
+    }
+
+    graph->forward();
+
+    return {data, std::move(graph), std::move(scorers)};
+}
+
+const TrlModel* trl_create_model(const char* yaml_config, const char* model, const size_t model_size, const char* source_vocab, const size_t source_vocab_size, const char* target_vocab, const size_t target_vocab_size, const char* short_list, const size_t short_list_size) {
```

Comment:
I'm assuming this is the start of the C implementation for things, if so shouldn't there be an extern "C"?

Reply:
It is in the header file 🙂